pandas/core/groupby/groupby.py

```diff
@@ -969,22 +969,16 @@ def reset_identity(values):
             result = concat(values, axis=self.axis)
             ax = self._selected_obj._get_axis(self.axis)
 
-            if isinstance(result, Series):
-                result = result.reindex(ax)
+            # this is a very unfortunate situation
+            # we can't use reindex to restore the original order
+            # when the ax has duplicates
+            # so we resort to this
+            # GH 14776, 30667
+            if ax.has_duplicates:
```

Comment:
we don't actually need to unique here, you can use index.get_index_non_unique which preserves ordering

Reply:
I have changed `get_indexer_for` to `get_indexer_non_unique` which is slightly faster. However, they essentially do the same thing (`get_indexer_for` calls `get_indexer_non_unique`) so we still need to take the unique values, otherwise there would be repeated data.